Ensure that calls can be described when ToString() throws

adamralph opened the pull request:
fixes #457

thomaslevesque commented on src/FakeItEasy/Expressions/ArgumentConstraints/EqualityArgumentConstraint.cs:
```diff
@@ -32,7 +33,14 @@ public override string ToString()
+            }
+            catch (Exception)
+            {
+                return this.ExpectedValue.GetType().Name;
```
Shouldn't the type be prefixed by "Faked" if `ExpectedValue` is a fake? This is what the default `ToString()` implementation does for fakes.

adamralph replied:
![image](https://cloud.githubusercontent.com/assets/677704/23332323/2819f990-fb78-11e6-8afd-d53e2aa0cf23.png)


But I do know that we had no test enforcing it! I'll add one, and alter the fix.

thomaslevesque replied:
👍

thomaslevesque replied:
In fact, [it *is* covered in tests](https://github.com/FakeItEasy/FakeItEasy/blob/376ba52f4126ea47e2bcaa494348612326714e4a/tests/FakeItEasy.Tests/Core/FakeManagerTests.cs#L335)... not in specs, though.

thomaslevesque replied:
In fact, [it *is* covered in unit tests](https://github.com/FakeItEasy/FakeItEasy/blob/376ba52f4126ea47e2bcaa494348612326714e4a/tests/FakeItEasy.Tests/Core/FakeManagerTests.cs#L335)... But adding a spec for this wouldn't hurt.

adamralph commented:
@thomaslevesque I inserted a tech-debt pay off commit at the beginning to enforce the "Faked... " convention in the messages via a spec.

I replaced the red commit completely and I added a fixup! commit for the green commit to special case fakes.

blairconrad commented on src/FakeItEasy/Expressions/ArgumentConstraints/EqualityArgumentConstraint.cs:
```diff
@@ -32,7 +36,18 @@ public override string ToString()
+                var type = this.ExpectedValue.GetType();
+
+                return type.Namespace == "Castle.Proxies"
+                    ? $"Faked {type.GetTypeInfo().BaseType.FullName}"
```
For a faked interface, this will print "Faked System.Object", rather than the faked interface.
And it misses faked delegates altogether.

Consider trying to get the FakeManager (Unfortunately we currently only have a `GetFakeManager` that throws when we call that on a non-Fake, as far as I can tell. But I see no reason not to expose an internal accessor that isn't so cranky.)
If `ExpectedValue` has a FakeManager, we can get the faked type from `FakeObjectType` and slap a "Fake " on the front of it.

While we're at it, the specs should probably include tests for faked interfaces, classes, and delegates, just to be sure we catch all the wacky cases.

adamralph replied:
OK, I'll give the `FakeManager` approach a try, and I'll add specs for the other types of types.

blairconrad commented on src/FakeItEasy/Expressions/ArgumentConstraints/EqualityArgumentConstraint.cs:
```diff
@@ -32,7 +36,18 @@ public override string ToString()
+
+                return type.Namespace == "Castle.Proxies"
+                    ? $"Faked {type.GetTypeInfo().BaseType.FullName}"
+                    : type.Name;
```
Why `Name` here and `FullName` above?

adamralph replied:
For some reason FakeItEasy uses the full name for fakes, whereas non-fakes only use the name:

```C#
Console.WriteLine(new List<int>().ToString());
```
> System.Collections.Generic.List\`1[System.Int32]

```C#
Console.WriteLine(A.Fake<List<int>>().ToString());
```

> Faked System.Collections.Generic.List\`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]

blairconrad replied:
Non-Fakes isn't quite just the name. It was the lack of namespace that tipped me off:

`System.Console.WriteLine((new List<int>()).GetType().Name);`

List`1

> System.Console.WriteLine((new List<int>()).GetType().FullName);

System.Collections.Generic.List`1[[System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]]

So just outputting `Name` still doesn't match the `ToString`, but it's perhaps a reasonable compromise.

The deeper issue, I suppose, why Fakes describe themselves differently. Maybe they shouldn't. Probably it doesn't matter. Either way, we don't need to address that question in this issue.

Thanks!

thomaslevesque commented on src/FakeItEasy/Expressions/ArgumentConstraints/EqualityArgumentConstraint.cs:
```diff
@@ -32,7 +33,26 @@ public override string ToString()
+                catch (ArgumentException)
+                {
+                    manager = null;
+                }
```
For now, the try/catch will do, but eventually we'll probably want to add a `Fake.IsFake` method and/or a `Fake.TryGetFakeManager` method. There's no point in throwing and catching an exception internally when it can easily be avoided.

adamralph replied:
All true. I briefly investigated, but I felt like I was going down a rabbit hole so I backed out.

blairconrad replied:
I'd imagined a `TryGetFakeManager` when I made the comment. I've added one in a commit that I've [pushed to my own branch](https://github.com/blairconrad/FakeItEasy/commit/4c9c637ee35cbd2d46b1f75354f14446ad26dfae). If you like it, feel free to squash it in.
If not, I'm very happy with what you have here. I particularly like the tests. Thanks!

adamralph commented:
I'll squash the fixup! commits after @blairconrad has given the diff his blessing (since he also chimed in above).

adamralph commented:
@blairconrad thanks for the fixup! commits. I've squash them into this branch. This should be ready to go now. /cc @thomaslevesque 

BTW - I was a bit unsure whether I preferred `bool TryGetFakeManager(object proxy, out FakeManager manager)` or `FakeManager TryGetFakeManager(object proxy)` but since you both seemed to prefer the latter, and since it's internal, I didn't agonise too long about it. If it were a public API I could see an interesting discussion arising.

blairconrad commented:
> BTW - I was a bit unsure whether I preferred `bool TryGetFakeManager(object proxy, out FakeManager manager)` or `FakeManager TryGetFakeManager(object proxy)`

I'm on the fence as well. I just didn't want to make a fuss. As you say, since it's internal, I also didn't agonise over it, and @thomaslevesque liked it, so I thought "why not?".

I also thought "in case there's controversy, I'll change the signature in an additional fixup, so @adamralph can decide whether to take 0, 1, or 2 of the changes".

adamralph commented:
> I'll change the signature in an additional fixup, so @adamralph can decide whether to take 0, 1, or 2 of the changes

I enjoyed that, and I almost took just the first one. 😉

blairconrad commented:
I believe @thomaslevesque already approved your base and the extra commits, so I'm-a merge.

Thanks, @adamralph!
